### PR TITLE
`Exam mode`: Generate student exam on demand if student is registered for the exam

### LINKED ISSUE #3835: Exams: Add Participants after Exam start
### Is your feature request related to a problem?

Esp. for long-time exams (longer than just a few hours), It may be important to add students after the exam itself has been started.

### Describe the solution you'd like

Possibility to add students (that work already) and generate individual exams for them (that don't work after exam start)

### Describe alternatives you've considered

_No response_

### Additional context

Student added, but generation of exam not possible:
image

Student's view that says he has to ask me (but I can't do anything)
image

### PR DESCRIPTION
**Pull request by coolchock**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

Currently, student exams can only be generated on demand for test exams, but not for real exams. This limitation can be problematic if additional students are registered for the exam and the instructor forgets to click `Generate missing student exams`. If this oversight is discovered after the exam has started, the only way to allow the student to participate is to duplicate the exam and register the missing students for the new one. This pull request adds a feature to generate student exams on demand for real exams as well.
resolves #3835 

### Description

The condition for allowing a student to generate a student exam (previously applicable only to test exams) has been extended to include a check to see if the student is registered for the exam. Now, a student exam will be generated if the exam is a test exam OR if the student is registered for the exam. This is possible only starting from 5 minutes before the exam to prevent inconsistencies, such as an instructor adding another exercise group after a student has generated their exam on demand.

Furthermore, if a student is registered for the exam after the exam has started, a student exam is now generated automatically.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam

1. Log in to Artemis
2. Navigate to Course Administration -> Exam Management as an instructor.
3. Register the first student for the exam before the exam start date, but do not generate the student exam.
4. Set the exam visibility date and exam start time so that the exam page is accessible to students more than 5 minutes before the exam starts.
5. Access the exam page as a student more than 5 minutes before the exam start time and verify that you see an error stating you are not registered for the exam.
6. Wait until there are less than 5 minutes before the exam start time, go to the exam page again, and verify that no error is displayed and that a student exam is generated. Also, verify this as an instructor under the "Student Exams" section in Exam Management.
7. Return to the instructor account and register the second student for the exam after the exam has started.
8. As an instructor, verify under the "Student Exams" section in Exam Management that the second student's exam has been generated.
9. As the second student, verify that you can participate in the exam.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
  - Enhanced flexibility for generating student exams based on registration status, allowing access for both test and normal exams.
  - Improved student exam registration process to accommodate late registrations, ensuring all students can access necessary exam materials.

- **Bug Fixes**
  - Improved clarity of error messages related to student registration for exams, making it easier for users to understand registration issues.

- **Tests**
  - Added new test cases to cover various user registration scenarios for exam access.
  - Enhanced integration tests to verify the correct generation of student exams upon registration.

### CODE REVIEW

**Comment by dfuchss** [comment]
Does this resolve https://github.com/ls1intum/Artemis/issues/3835 as well?

**Comment by coolchock** [comment]
@dfuchss yes, it does. Students can be added to an exam after it has started, and their individual student exams will be generated on demand. It's also possible to adjust individual working time.

**Review by dfuchss — APPROVED** [review]
Code

**Review by sarpsahinalp — APPROVED** [review]
Code lgtm

**Review by pzdr7 — APPROVED** [review]
Worked for me on TS2.

----

Potential edge case: If you add an exercise group after a student generates a student exam this way, the student exam is inconsistent with the exam.

(Follow up) Should we prevent such changes once the exam is visible?

**Review by magaupp — APPROVED** [review]
Tested on TS5 and locally. The exams were created without errors.

**Review by SimonEntholzer — APPROVED** [review]
Code

File: src/test/java/de/tum/in/www1/artemis/exam/ExamRegistrationIntegrationTest.java
Comment (JohannesStoehr) [inline src/test/java/de/tum/in/www1/artemis/exam/ExamRegistrationIntegrationTest.java:130]: This test seams to fail

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/test/java/de/tum/in/www1/artemis/exam/ExamRegistrationIntegrationTest.java
Comment (coolchock) [reply src/test/java/de/tum/in/www1/artemis/exam/ExamRegistrationIntegrationTest.java:130]: fixed

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamAccessService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/exam/ExamAccessService.java:89]: Should we also check if the exam is not yet over? Otherwise we might get weird cases where a student can create an exam after it is finished

**Review by SimonEntholzer — DISMISSED** [review]
Works on TS4 :+1:

**Review by florian-glombik — APPROVED** [review]
Tested on ts4, registered a new student to a running exam and the student exam was generated right away

**Review by EneaGore — APPROVED** [review]
Testing Session. Worked on TS4.

**Review by sarpsahinalp — DISMISSED** [review]
Testing Session. Worked on TS4.

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamAccessService.java
Comment (coolchock) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamAccessService.java:89]: done

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by SimonEntholzer — APPROVED** [review]
Still works on Ts3 :+1:

**Review by EneaGore — DISMISSED** [review]
re-approve after merge conflicts

**Review by magaupp — APPROVED** [review]
Works on TS5 :+1:

**Review by pzdr7 — APPROVED** [review]
Reapprove after merge conflict

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge conflicts

**Review by dfuchss — DISMISSED** [review]
Re-Approve

**Review by SimonEntholzer — APPROVED** [review]
reapprove

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge conflicts

**Review by krusche — APPROVED** [review]
Code looks good to me 👍

**Pull request merged**
